add FWaaS extension to openstack-neutron

andreaturli opened the pull request:
@zack-shoylev could you please review it?

I'd like to merge that on master and 1.9.x if you think it is ok

LiveTests are ok for me, but `removeFirewallRuleFromPolicy` as it seems that the [doc](https://docs.google.com/document/d/1PJaKvsX2MzMRlLGfR0fBkrMraHYF0flvl0sqyZ704tA/edit#) is not up-to-date

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/Firewall.java:
```diff
@@ -0,0 +1,207 @@
+      this.firewallPolicyId = firewallPolicyId;
+   }
+
+   private Firewall() {}
```
Use autovalue with builders to eliminate most of the boilerplate.

andreaturli replied:
I think all of the domain objects need to be refactored with AutoValue. Do we want to start from these new ones?

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/FirewallRule.java:
```diff
@@ -0,0 +1,363 @@
+
+public class FirewallRule {
+
+   public enum IpVersion {
```
Consider using a separate file for the enum. It will be more readable.

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/extensions/FWaaSApi.java:
```diff
@@ -0,0 +1,250 @@
+   @ResponseParser(ParseFirewallRules.class)
+   @Fallback(EmptyFirewallsFallback.class)
+   @Path("/firewall_rules")
+   FirewallRules listFirewallRules(PaginationOptions options);
```
Is it possible to switch this to using PaginatedCollection instead, similar to 
https://github.com/jclouds/jclouds-labs-openstack/blob/master/openstack-poppy/src/main/java/org/jclouds/openstack/poppy/v1/features/ServiceApi.java#L92

That way, you should be able to eliminate the FirewallRules class, and might be able to aply to a few more classes.

andreaturli replied:
thanks for the suggestion

zack-shoylev commented on openstack-neutron/src/test/java/org/jclouds/openstack/neutron/v2/extensions/SecurityGroupApiLiveTest.java:
```diff
@@ -47,6 +48,10 @@ public void testCreateUpdateAndDeleteSecurityGroup() {
             sgApi = api.getSecurityGroupApi(region).get();
 
+            for (Subnet subnet : api.getSubnetApi(region).list().concat().toList()) {
+               System.out.println("subnet: " + subnet.getName());
```
?

andreaturli replied:
ops, sorry about that

andreaturli commented:
Thanks @zack-shoylev 
I've addressed your comments:
- I've added AutoValue for Firewall, FirewallPolicy and FirewallPolicyRule 
- Add PaginatedCollection where possible
and some other minor fixes

Does it look better now?

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/Firewall.java:
```diff
@@ -0,0 +1,69 @@
+   @Nullable public abstract String getTenantId();
+   @Nullable public abstract String getName();
+   @Nullable public abstract String getDescription();
+   public abstract boolean isAdminStateUp();
```
Might want to consider using Boolean instead, if this is not a required parameter.

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/FirewallPolicy.java:
```diff
@@ -0,0 +1,81 @@
+   @Nullable public abstract List<String> getFirewallRules();
+   public abstract boolean isAudited();
+
+   FirewallPolicy() {
```
I don't think this is needed at all?

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/FirewallPolicy.java:
```diff
@@ -0,0 +1,81 @@
+@AutoValue
+public abstract class FirewallPolicy {
+
+   @Nullable
```
Formatting

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/FirewallPolicy.java:
```diff
@@ -0,0 +1,81 @@
+   @Nullable public abstract String getTenantId();
+   @Nullable public abstract String getName();
+   @Nullable public abstract String getDescription();
+   public abstract boolean isShared();
```
@Nullable Boolean?

andreaturli replied:
I think I've managed required shared parameter at line 57

zack-shoylev replied:
I see it, just wanted to make sure actually required by the service.

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/extensions/FWaaSApi.java:
```diff
@@ -0,0 +1,248 @@
+   @POST
+   @SelectJson("firewall")
+   @Path("/firewalls")
+   Firewall create(@WrapWith("firewall") Firewall firewall);
```
While this is possible and very simple/effective, it is discouraged. The preferred way to do this might be
```
Firewall create(@WrapWith("firewall") CreateFirewall firewall);
```
Basically, we should provide slightly different CreateFirewall and UpdateFirewall classes - for example, they don't support specifying IDs or other parameters that are strictly determined by the service (another example is Status).

There is some repetition, yes, but we have autovalue to take care of that.

andreaturli replied:
Are you thinking about something like `openstack-poppy's Service` and `openstack-poppy's CreateService` ? I'm not sure I like it but I think it is better than plain `Firewall`. 
Maybe @nacx has another suggestion?

zack-shoylev replied:
That's exactly right - similar to poppy.

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/extensions/FWaaSApi.java:
```diff
@@ -0,0 +1,248 @@
+   @Nullable
+   FirewallPolicy removeFirewallRuleFromPolicy(@PathParam("id") String policyId, @WrapWith("firewall_rule_id") String firewallRuleId);
+
+
```
Extra line

zack-shoylev commented:
Thanks @andreaturli ; looking pretty good

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/CreateFirewallPolicy.java:
```diff
@@ -0,0 +1,75 @@
+   }
+
+   public static Builder builder() {
+      return new AutoValue_CreateFirewallPolicy.Builder();
```
I am guessing your builder needs to have isShared and isAudited set to default values (false) here.

Also based on this:
http://docs.openstack.org/admin-guide-cloud/content/fwaas_api_abstractions.html

Please add the corresponding docs for the properties here and in other places as applicable.

zack-shoylev replied:
Haven't tried it, but this might fix it:
```
return new AutoValue_CreateFirewallPolicy.Builder().shared(false).audited(false);
```

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/Firewall.java:
```diff
@@ -0,0 +1,66 @@
+
+   public abstract Builder toBuilder();
+
+   @AutoValue.Builder
```
Only CreateFirewall and UpdateFirewall will need a builder, you can switch to a regular autovalue class here (with no builder).

zack-shoylev commented:
Another thing I overlooked - openstack-heat has the good up-to-date examples that I recently refactored.

zack-shoylev commented:
Also I think the getSomeName() getters were fine. Auto works fine with
getName()
and name(String name)/getName() in the builder, for example.

https://github.com/jclouds/jclouds-labs-openstack/blob/master/openstack-heat/src/main/java/org/jclouds/openstack/heat/v1/options/CreateStack.java

andreaturli commented:
thanks @zack-shoylev for your help. For some reason, when using getSomeName() format the builder was not generated at all. I think we can solve this issue later on, as other domain objects need to be refactored with AutoValue, anyway. Wdyt?

zack-shoylev commented:
That is strange. Sure, but maybe have a look and see it's not something else that's going wrong.

zack-shoylev commented on openstack-neutron/src/main/java/org/jclouds/openstack/neutron/v2/domain/CreateFirewall.java:
```diff
@@ -0,0 +1,123 @@
+      abstract CreateFirewall autoBuild();
+
+      public CreateFirewall build() {
+         CreateFirewall createFirewall = autoBuild();
```
Eh, that's messy.
It might be better to implement the builder getters. That way users benefit from having them; and you benefit from being able to check validation before sending this to the service;

As an afterthought - I don't really like client-side validation of parameters as is done here. Let the service validate.

andreaturli commented:
@zack-shoylev thanks again. I think now the AutoValue classes are in a better shape

andreaturli commented:
I'd appreciate if you could merge it on both master and 1.9.x. Thanks!

zack-shoylev commented:
squashed, merged, backported

andreaturli commented:
thanks @zack-shoylev
